### PR TITLE
Pass torch.distributed launch process local rank as environment variable instead of argument 

### PR DESCRIPTION
**Pull request by jwwandy**
In `torch.distributed.launch.py`, it passes `local_rank` as argument and requires user's program to parse it. However, it would be more flexible for users and consistent with other variables, e.g. `RANK`, `MASTER_PORT`, `WORLD_SIZE`, if passing through environment variables.

https://github.com/pytorch/pytorch/blob/265ed8ff451c17eac82050b1767837ec924d9591/torch/distributed/launch.py#L200-L212

### CODE REVIEW

**Comment by pietern** [comment]
Thanks for submitting a PR :)

This will break existing users, so we shouldn't merge this as is. All environment variables you mention don't have to be explicitly mentioned by the script to run. If you run `torch.distributed.init_process_group` it picks all of them up by default.

**Comment by jwwandy** [comment]
@pietern I understood it would break the current user. But in my opinion, passing through a argument limits the usability of this launcher. For example, if using some [click](https://click.palletsprojects.com/en/7.x/commands/#multi-command-chaining) package multi-command pipeline. The user should always parse local rank as last argument in current scenario and this might not always be the best. I'm thinking of if there's any solution that is backward compatible but also fix this issue.

**Comment by apaszke** [comment]
We could add a flag like `use_env=False` that you could use to enable this mode.

**Comment by fmassa** [comment]
Question: what was the original motivation for adding the `--local-rank` to the main script? Just to avoid querying from an env var? Or is it more fragile?

**Comment by pietern** [comment]
+1 on the `use_env` flag. We could choose to make this the default the next major release to simplify things, but seeing as use of this tool is pervasive I wouldn't want to break anybody using it in the 1.0 versions.

The local rank is needed to make sure each locally launched process uses the right device. Either you pass the local rank down, or pass the number of processes per node down and derive it yourself.

It was added in #5348 and looking at the PR there was no discussion around using an environment variable or an argument for this parameter.  @teng-li Do you recall?

**Comment by pietern** [comment]
@jwwandy The behavior should be such that existing users don't have to do anything, so `use_env=False` as default. Then you can pass `--use-env=1` to the launch utility to enable it and stop the tool from passing down the additional argument. Then once this is merged we can make an issue to change this behavior for the next major release.

**Comment by jwwandy** [comment]
@pietern Sounds great. I've pushed a new version for this pull request. Please help check if anything missed. Thanks.

File: torch/distributed/launch.py [hunk 1]
```diff
@@ -170,6 +183,11 @@ def parse_args():
                         help="Master node (rank 0)'s free port that needs to "
                              "be used for communciation during distributed "
                              "training")
+    parser.add_argument("--use_env", default=False, type=bool,
```
[the remark this answers is not in the document]
Comment (jwwandy) [reply torch/distributed/launch.py]: Got it . Thanks.

**Pull request closed without merging**